src/Build.UnitTests/ChangeWaves_Tests.cs

```diff
@@ -0,0 +1,219 @@
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                collection.LoadProject(file.Path).Build().ShouldBeTrue();
+                log.AssertLogContains("Hello World!");
+            }
+        }
+
+        [Theory]
+        [InlineData("test")]
+        [InlineData("    ")]
+        [InlineData("")]
+        [InlineData("16-7")]
+        [InlineData("16x7")]
+        [InlineData("16=7")]
+        public void InvalidCallerForIsFeatureEnabledThrows(string waveToCheck)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", "16.8");
+                Shouldly.Should.Throw(() => ChangeWaves.IsChangeWaveEnabled(waveToCheck), typeof(InternalErrorException));
+            }
+        }
+
+        [Theory]
+        [InlineData("test", "16.8")]
+        [InlineData("16_8", "5.7")]
+        [InlineData("16x8", "20.4")]
+        [InlineData("garbage", "18.20")]
+        public void InvalidFormatThrowsWarningAndLeavesFeaturesEnabled(string enabledWave, string featureWave)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", enabledWave);
+                ChangeWaves.IsChangeWaveEnabled(featureWave).ShouldBe(true);
+
+                string projectFile = @"
+                    <Project>
+                        <Target Name='HelloWorld' Condition=""'$(MSBUILDDISABLEFEATURESFROMVERSION)' == '" + ChangeWaves.EnableAllFeatures + @"' and $([MSBuild]::VersionLessThan('" + featureWave + @"', '$(MSBUILDDISABLEFEATURESFROMVERSION)'))"">
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                Project p = collection.LoadProject(file.Path);
+                p.Build().ShouldBeTrue();
+
+                log.AssertLogContains("invalid format");
+                log.AssertLogContains("Hello World!");
+            }
+        }
+
+        [Theory]
+        [InlineData("0.8")]
+        [InlineData("203.45")]
+        public void OutOfRotationWavesThrowsWarningAndDisablesFeatures(string enabledWave)
```

Comment:
What's supposed to happen if I specify something between two versions?

Reply:
Say the waves are 16.8, 16.10, and 17.0 and you specify 16.10:
16.8 will be enabled
16.10 will be disabled
17.0 will be disabled.